### ISSUE TITLE
Add Serverkess optimize plugin

### ISSUE THREAD

**Issue by pmuens** [serverless/serverless#1653]
This plugin should make it able to optimize the functions code (e.g. minify) before deployment.

**Comment by flomotlik** [comment]
Do we need a separate plugin for that, or would it be possible to do this via build tools that people use? Imho great documentation on this would go a long way of setting this up in a nice way, but maybe people would prefer this in a plugin. @serverless/vip any thoughts on that?

**Comment by ajagnanan** [comment]
The optimize plugin worked really well for me for reducing function size. Yea, the browserify issues are there, but just gotta work around them. I think its pretty much a must have on any serverless project. I would rather have it as a plugin than setting build tools up to do it.

**Comment by rajington** [comment]
I think traditionally documentation could be sufficient, but I feel serverless has taken the role (and may contribute some of its success) to being a "low-decision-overhead, pattern-filled" framework vs. a lightweight library. Past users might be angry that "v1 has less features than v0.5 does", and new users might not want to be burdened with how to set that up in addition to everything else.

There's also things like "tree shaking" in rollup/Webpack 2 that is perfect for serverless, but users would have to stay on top of those types of things to note the documentation change, vs a plugin getting a new version that supports it.

Maybe this is more thought than you were looking for, but I've seen ARGUABLY "better products" (Flow/Babel) be dismissed in favor of ARGUABLY "worse products" (TypeScript) because of it was easier on devs. I also wouldn't underestimate path dependence when it comes to getting devs to upgrade their projects to v1.

The downside is of course having to support it...

**Comment by kennu** [comment]
I agree with @rajington. To me Serverless is all about simplifying the developer experience of "raw" Lambda/API Gateway. I can easily generate a project and it will just work, without worrying about setting up a complicated build environment and connecting bits and pieces manually. Also any developers joining my project will immediately understand the "Serverless way", without having to learn a customized build setup.

Historically I would compare this to e.g. Ruby on Rails, which very nicely generates a working project with all bells and whistles included. Or to the Angular-Fullstack Yeoman generator, which did the same for Angular/Express/Grunt/Gulp projects. They are arguably bloated tools, but they provide a very productive developer experience, which doesn't require you to understand everything that's happening under the hood.

**Comment by shortjared** [comment]
I would argue a plugin would be useful, but not necessary until after we have something stable.

However, with the Alpha we are having pretty good success with just an `npm build` that looks something like `npm run flow && babel src -d lib --copy-files`. The Alpha brings a level of simplicity to the project that wasn't afforded by the typical `s-function.json`. Granted there is a non-trivial amount of configuration and setup to get to that point alone. Sticking on rollup or WebPack 2 should be fairly straightforward then.

**Comment by andrew-templeton** [comment]
Requirements first look at it...
- I need to be able to NOT minify for debugging, on any deployment to any environment or platform
- I need to be able to instantly start building with the seed/core scaffolded default project configuration if I don't really know about minning or why I should do it
- I may want to be able to override the optimize/min step(s) and add my own optimize step(s) if I (1) am aware of what/why (2) have a reason to change behavior
- I should NEVER have to work against the tool's default configuration, or have "gotchas" that do not work one-to-one with my un-minned code in the default configuration

My suggestion:
- Make it a default plugin that is OFF for anything below staging/production and ON for staging and production
- Use a tool that has zero gotchas, even if suboptimal (for example, don't elect to make something like closure compiler advanced shaking which might break certain type conversions) for the default configuration
- Allow a well-defined and well-documented method to author a replacement plugin step, for advanced users

From a user's perspective, we have two buckets:
- If I have a well-understood optimization process for my code elsewhere already and I want to apply this to Serverless, I am then also guaranteed to be advanced enough to shim my own step in with a good interface, to do the override.
- If I lack this process for my code elsewhere, I am going to get a lot of benefit from having this problem solved for me with no extra steps.
- In both cases, I am not looking for Serverless to solve this problem for me as a primary issue. Either (1) I don't know I need it and I'm going to just be delighted to have it for free with default, less-optimal config, or (2) I'm going to be advanced, aware I need to override to my requirements, and I'm using Serverless for the other problems it solves.
- The gray-area case is that I'm aware of these things and that I should be optimizing but don't practice them now. I then will be grateful that a reasonable default configuration that doesn't have gotchas exist. I am unlikely to be *dissatisfied* that the solution is using a theoretically slightly less optimal tool.


Sounds like it should be a plugin, but a plugin that's in the default set that's turned on, with smart and simple defaults that have no complicating or leaky side-effects.

**Comment by pmuens** [comment]
Thanks for the great feedback so far!

I agree that Serverless should be a help for the developer to get things up and running easily.

How do you guys think about having a plugin (which may lack flexibility) vs. setting up everything manually once. Is it a tedious process to setup everything or might be a good documentation how one can do it sufficient instead of writing a plugin?

Additionally it would be interesting if some of you has statistics / metrics regarding unoptimized vs. optimized functions (e.g. size of function, execution time, deployment time etc.).

Thanks in advance!

**Comment by HyperBrain** [comment]
Currently we use the node4.3 runtime everywhere and also have our submodule dependencies converted to ES6. Minify has to be disabled completely, as the currently used `uglify` does not handle ES6 dependencies correctly. Additionally it is IMHO mandatory to have a selection option for the babel profile to be used (currently we use ES2015-node4).
For debugging and test builds there has to be an easy way to disable the plugin completely.
Forgot to say: Personally I would prefer the optimizer as plugin.
Just my 5 cents.

**Comment by flomotlik** [comment]
Been thinking a lot about this all day today, here my 2 cents. 

Basically I see a few issues with implementing Optimisation in Serverless at this point:

1. Optimisation works really well and out of the box for small projects or teams with small and not very complex codebases as those will most likely not run into any issues with the minifier/uglifyer. Those projects are on the flip side also the ones where optimisation has the least impact as they are deployed less often (smaller team) and have fewer function runs (as they are a smaller project)

2. Teams that benefit the most from optimisations will in my opinion want more control over the exact process that happens during minification. e.g. @HyperBrain mentioned issues with uglify. My Background is in build processes (at Codeship) and most larger teams I've come across will want to be very specific about optimisations in parts of their builds where it really matters and has a large impact.

3. I don't have good data yet what the impact of minification is on deployment. As we're zipping the handler files and include other dependencies as well a zipped file should be compressed down to a similar file size than running it through a minifier before (is my assumption, haven't tested it yet). It might not be the same, but imho good compression algorithms should get it to close enough to not matter. Would be interesting to get more data from your projects, e.g. @shortjared what is the size of the deployed zip when you run minify beforehand and when you don't?

4. I also don't have a good enough understanding of the impact of it on Cold start times. I don't doubt that it does have an impact, but how often do cold starts really happen for functions compared to total executions and if the cold start time optimisation is something a team wants to do a lot for could we do a good enough job that they would use our tooling instead of simply implementing their own optimisations that they could integrate into the build and might only take a few minutes to set up? If you guys either have better numbers on current cold start times ore are interested in working with us to get those numbers out of your production systems I'd be super interested.

5. NPM dependencies aren't necessarily minified, so if we minify the code that is provided by users, but don't minify the code provided through NPM does it really make that much of a difference?

6. With all of those points do we want to spend a lot of time (and imho getting it right and working out of the box is quite a bit of work) on this or aren't there simply a lot higher priority tasks that we can cover and simply provide great documentation about optimising your functions that teams who have the need for optimisation can use.

**Comment by HyperBrain** [comment]
7. Optimizer even without minify is also used to reduce the uploaded code that is part of a bigger (shared) codebase to the needed components only. Otherwise e.g. mySQL or other big components would be included in each and every function whether they use it or not. In practice that led to very minimal code sizes (even without minify) depending on the function's real use of the code.

**Comment by flomotlik** [comment]
@HyperBrain how did the optimizer plugin do this? At least in here I didn't find anything: https://github.com/serverless/serverless-optimizer-plugin

I do fully agree though that being able to craft minimal zip files is important, but we should improve the packaging mechanism there, so doesn't really need a separate optimization plugin or step then

**Comment by HyperBrain** [comment]
@flomotlik 
With this configuration the plugin creates one serverless_handler.js file out of the sources, but does not apply minify/uglify. The not used sources are kept out.
```
  "custom": {
    "excludePatterns": [],
    "optimize": {
      "exclude": [
        "aws-sdk",
        "mysql"
      ],
      "transforms": [
        {
          "name": "babelify",
          "opts": {
            "presets": [
              "es2015-node4"
            ]
          }
        }
      ],
      "includePaths": [
        "config",
        "node_modules/mysql"
      ],
      "minify": false
    }
  },
```

**Comment by wei-xu-myob** [comment]
I'm using FlowType & async in lambda, and using Babelify to transform under Serverless 0.5.6.


I want to know how to deal with the custom package processes in 1.0, I could not find the extension point from the source code. 


Any suggestion for this?

**Comment by pmuens** [comment]
Hey @wei-xu-myob thanks for reaching out.

The `package` plugin currently hooks into the `deploy:createDeploymentPackage` lifecycle. You can simply hook into the `before:deploy:createDeploymentPackage` lifecycle to deploy your assets before they are zipped and deployed.

You can do this with the help of a plugin. You can find the docs on how to develop a plugin here: https://github.com/serverless/serverless/blob/master/docs/developing-plugins/building-plugins.md

Let me know if you need any help!

**Comment by pmuens** [comment]
Just want to note that @thenikso wrote a great webpack plugin you can now use with Serverless v1:
https://github.com/elastic-coders/serverless-webpack

Here's the open issue in our `community-plugins` repository:
https://github.com/serverless/community-plugins/issues/2

**Comment by wei-xu-myob** [comment]
Note to my previous request, I wrote a plugin to support execute a js script at any event.

For example, to run a customized script to package file before deploy:

```
plugins:
  - serverless-scriptable-plugin

custom:
  scriptHooks:
    before:deploy:createDeploymentPackage: build/package.js
```

I'm using Babel to transcompiling and zip the required runtime dependencies in my customized package script.

Examples provided in the source code:
https://github.com/wei-xu-myob/serverless-scriptable-plugin

**Comment by andymac4182** [comment]
I think leaving this as a plugin for now works really well as there are so many different ways and tools people are using. It would be great to see a list of recommended methods in the documentation. This would allow people who are already doing this to know how to get it working within the framework and give people a start if they haven't done this before. 

It would also be good if there was some additional hooks for when doing something like `watch` which is currently handled completely by plugins. We use `serverless webpack serve` from `serverless-webpack` to do local testing and basic debugging but the `serverless-offline` plugin has a nicer local debugging experience. It would be good if there was a `watch` or `serve` plugin architecture that they could integrate with. That would also resolve some of the offline support discussed in https://github.com/serverless/serverless/issues/1582

**Comment by j** [comment]
Ugh.  Why doesn't serverless just have certain hook options that can be used for each function step like Apex?  Working with webpack in apex is so easy.. That way each function can be built and bundled on it's own.

I'd be a happy man if I just had:

```yml
functions:
  hello:
    handler: build/hello/handler
    hooks:
      - build: webpack --config ./handlers/hello/webpack.config.js
      - clean: rm -rf build/hello
  unrelated:
    handler: build/unrelated/handler
    hooks:
      - build:
        - babel-node bin/run deploy:assets
        - babel-node bin/run build
      - clean: rm -rf build/unrelated
```

Something without maintaining bloated plugins from random people.  serverless-webpack is a cool option for people, but there's already been 3 instances where something is broken or not working with it or unoptimized.

**Comment by nicka** [comment]
@j I believe adding this as "standard" functionality within the framework makes a lot of sense. 

We've been using [serverless-plugin-hookscripts](https://github.com/kennu/serverless-plugin-hookscripts) a lot with `v0.5`. The alternative for `v1.0` [serverless-scriptable-plugin](https://github.com/wei-xu-myob/serverless-scriptable-plugin) would give you the same behaviour, but tbh having access to hooks should be possible without any plugin.

**Comment by j** [comment]
Yes, I completely agree with that. Especially on the configuration level.

Sometimes I don't want to write JS to build something that is as easy as typing a webpack build command and then a rm dir command.

I ended up straying away from serverless-webpack and now build before I deploy, but it would be cool to not have to do that. See 'serverless-graphql' library. They'd clean a lot of their complexity with build hooks. That code is a huge learning curve for a "boilerplate".

**Comment by johncmckim** [comment]
There seems to be three questions, should we optimise lambdas, how should we optimise lambdas and what's the user experience for this. Part of the issue is that the best way to optimise code might also depend on how we structure our code i.e. monolithic lambdas vs single purpose functions. I think babel gets mixed in here because of webpack, but I think transpilation is separate to optimization.

To the first question, I think the assumption is that reducing the package size will lead to better performance on Lambda. But, I think as @flomotlik points out we need better data on that. Reducing the package size does lead to faster deployments which is beneficial. But it depends on how we achieve that.

For me, the downside of minifying code on Lambda is the affect on logs. If there is an error thrown from minified code it is difficult to work out what that corresponds to in your un-minified code. You could only minify your code in production. But, you will have bugs in production and those are the most important bugs to fix quickly.

If we're creating single purpose functions, packaging functions individually goes a long way to reducing overall lambda size. As @rajington points out, add tree shaking on top of that and we could reduce the file size significantly without having to minify code. For me, unless there is really compelling data to say minification improves performance by x, I don't want to use it.


In terms of implementing this I think we could learn a lot from the React community. For me starting a react project as a beginner was a nightmare. I suddenly had to learn webpack + react. Even now, starting a new project with webpack + react, can be a nightmare. If I don't use a tool, I have to deal with so much boiler plate webpack config. It's a bad user experience and has given rise to tools like [create react app](https://github.com/facebookincubator/create-react-app) and [nwb](https://github.com/insin/nwb). Those tools are great because they make a lot of assumptions for me, but also allow me to drop in little bits of webpack config where I need to.

While build hooks are useful for some people. I don't think users should have to configure build config per function to optimise their code. I think that could lead to a bad user experience with a lot of boilerplate code.

**Comment by erikerikson** [comment]
I don't have hard data but it has been verified by Amazon that smaller packages reduce cold start latencies by reducing the number of bytes passing over the network into and being unpacked by the Lambda containers.

It should be easy to collect such data, if time consuming.

**Comment by nicka** [comment]
@johncmckim Thanks for the response, sorry I was mixing optimizing with hooks here a bit! While in fact that are two separate things.

My general though about this is, we should improve the packaging mechanism here. For me `include` and `exclude` works. But in some cases you want something prior happing before the `include/exclude`. In those scenarios I'd rather configure it inside the `serverless.yml` next to the actually impacted functions(with hooks). This would allow new devs to see what's happening from within one file instead of also taking in account a `package.json` or something similar.

Off course this is just an opinion and already possible with the `serverless-scriptable-plugin`, but I don't see why we don't want to offer this by default?

**Comment by flomotlik** [comment]
# On Minification and Code optimisation

I'm still not convinced that minification and other code optimisation techniques really help the deployment speed on AWS in any meaningful way.

Our current understanding is that Lambda has a 3 step process to start your functions:

1. Your Code is downloaded from S3 onto an EC2 instance to run
2. It starts a container to run your code inside of it
3. The Lambda management process loads your code and then executes a Lambda function

Now after that we have a hot function. When they retire the function it seems like the following steps are happening:

1. Stop the Lambda Management process and container after a bit
2. Remove the downloaded Code from the machine a longer time later

From this imho there are things that make sense to optimise and others that don't Removing unnecessary dev dependencies from your node_modules folder for example, or only including what is actually necessary to be deployed are great optimisations because they cut a lot of fat. Minifying your Code should have practically 0 influence on the start time of your function though as the minimal optimisation you get from that will make no difference considering how fast the internal network of AWS is and that this code only gets downloaded from machines relatively rarely if your functions run regularly enough.

# Why I think Serverless shouldn't include hooks or any calls to other tools

The main reason why I'm hesitant to include any kind of external command/hook support into serverless because from my experience of building CI/CD systems at Codeship and setting it up for many many projects mixing deployment and workflow tools will be a PITA very quickly.

A simple Node example would be using different PATH variables than e.g. NPM would do when calling a command and then you're using a globally installed version of something instead of the locally installed one. And then your packaging doesn't work like it should be and you have no idea why.

It also increases the complexity of our Code because dealing with 3rd party commands is always hard. How do you print the logs in a nice and consistent way, how do you support coloring of error messages, STDIN/STDOUT/STDERR support, good exit status, .. are all painful to do.

Especially considering that there are tons of really good tools out there to do all of this in preparation to the deployment of serverless. Through bash, make, gulp or many other tools its very easy to set up really good deployment processes where you can be very specific about dependencies between commands.

I want Serverless to be laser focused on being the best FaaS deployment and infrastructure management tool. There are other great build tools out there that do a fantastic job and I don't want to spend time half-competing with them when I could put more resources on building more important features that will make a huge difference for how everyone builds infrastructure on Lambda and other providers.

**Comment by j** [comment]
I don't really believe in the minification / code optimization aspect of this original question.

Personally, I hate trusting all these non battle-tested 3rd party "plugins" that have a wonky code-base (no offense to those contributors) that do something like run an asset bundling command or a webpack build command when those can just be one liner CLI command.

NPM has `npm run` which is where people's build commands typically go.  But doing something in the same webpack flow is a pain.  Apex just feels good when for having hooks built into the function config file there.  It's easy.

So really the only solution besides trusting 3rd party libraries is to do something like https://github.com/serverless/serverless-graphql.

**Comment by flomotlik** [comment]
@j yeah we have to find a balance around what we implement for usability, which features are important and what is a nice need that a plugin can fulfill. I do think some of the plugins are definitely production ready and we're going to work more with plugin developers in the future to make that better and have a clearer message to the community.

But yeah personally I also prefer Bash scripts or build tools, but I built CI services for the last 6 years at my last Startup so that is probably not fair to compare to other people :D.
